[JSC] Implement `Iterator.from` from Iterator Helpers Proposal

sosukesuzuki opened the pull request:
#### dab2b14ea3afc85a224fe7f83fe20938baeb0a60
<pre>
 [JSC] Implement `Iterator.from` from Iterator Helpers Proposal
<a href="https://bugs.webkit.org/show_bug.cgi?id=278691">https://bugs.webkit.org/show_bug.cgi?id=278691</a>

Reviewed by Yusuke Suzuki.

This patch implements `Iterator.from`[2][3] from the Iterator Helpers proposal[1].

[1]: <a href="https://github.com/tc39/proposal-iterator-helpers">https://github.com/tc39/proposal-iterator-helpers</a>
[2]: <a href="https://github.com/tc39/proposal-iterator-helpers?tab=readme-ov-file#iteratorfromobject">https://github.com/tc39/proposal-iterator-helpers?tab=readme-ov-file#iteratorfromobject</a>
[3]: <a href="https://tc39.es/proposal-iterator-helpers/#sec-iterator.from">https://tc39.es/proposal-iterator-helpers/#sec-iterator.from</a>

* JSTests/stress/iterator-from.js: Added.
(assert):
(shouldThrow):
(throw.new.Error.Iter.prototype.next):
(throw.new.Error.Iter):
(throw.new.Error):
(sameValue):
(sameValue.Iter.prototype.return):
(sameValue.Iter):
(sameValue.Iter.prototype.next):
(sameValue.const.iter.get next):
(sameValue.const.iter.get return):
(sameValue.):
(sameValue.get const):
(string_appeared_here.shouldThrow):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/JSIteratorConstructor.js: Added.
(linkTimeConstant.getIteratorDirect):
(linkTimeConstant.getIteratorFlattenable):
(from):
* Source/JavaScriptCore/builtins/WrapForValidIteratorPrototype.js: Added.
(next):
(return):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
(JSC::BytecodeGenerator::emitIsWrapForValidIterator):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/heap/HeapSubspaceTypes.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::wrapForValidIteratorPrototype const):
(JSC::JSGlobalObject::wrapForValidIteratorStructure const):
* Source/JavaScriptCore/runtime/JSIteratorConstructor.cpp:
(JSC::JSIteratorConstructor::create):
(JSC::JSIteratorConstructor::finishCreation):
* Source/JavaScriptCore/runtime/JSIteratorConstructor.h:
* Source/JavaScriptCore/runtime/JSType.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/runtime/JSType.h:
* Source/JavaScriptCore/runtime/JSWrapForValidIterator.cpp: Added.
(JSC::JSWrapForValidIterator::createWithInitialValues):
(JSC::JSWrapForValidIterator::finishCreation):
(JSC::JSWrapForValidIterator::visitChildrenImpl):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSWrapForValidIterator.h: Added.
* Source/JavaScriptCore/runtime/JSWrapForValidIteratorInlines.h:
(JSC::JSWrapForValidIterator::createStructure):
* Source/JavaScriptCore/runtime/WrapForValidIteratorPrototype.cpp:
(JSC::WrapForValidIteratorPrototype::finishCreation):
* Source/JavaScriptCore/runtime/WrapForValidIteratorPrototype.h:
* Source/JavaScriptCore/runtime/WrapForValidIteratorPrototypeInlines.h:
(JSC::WrapForValidIteratorPrototype::createStructure):

Canonical link: <a href="https://commits.webkit.org/283016@main">https://commits.webkit.org/283016@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/58c58824ad077635697f978269d88e1696933aef

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64934 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44301 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17548 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68958 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15540 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52084 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15822 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/52170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/10725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68000 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/40965 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/56194 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/32792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37636 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/13566 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14416 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/58047 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59545 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/13904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70663 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/64177 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8886 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13383 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/59498 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8918 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/56255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59704 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14333 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/7329 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/1000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/85943 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/40113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15148 "Found 4 jsc stress test failures: jsc-layout-tests.yaml/js/script-tests/function-apply-aliased.js.layout-no-cjit, stress/json-stringify-inspector-check.js.no-llint, wasm.yaml/wasm/stress/cc-int-to-int-cross-module-with-exception.js.default-wasm, wasm.yaml/wasm/stress/cc-int-to-int-no-jit.js.default-wasm") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/41190 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/42371 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40934 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/builtins/WrapForValidIteratorPrototype.js:
```diff
@@ -0,0 +1,79 @@
+{
+    "use strict";
+
+    @putByIdDirectPrivate(this, "iterated", iteratorRecord);
```
We would like to remove getDirect / putDirect things more and more. So let's implement it with JSInternalFieldObjectImpl instead.

sosukesuzuki replied:
I've added `JSWrapForValidIterator` that extends `JSInternalFieldObjectImpl`

Constellation commented on Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
```diff
@@ -0,0 +1,86 @@
+        iterator = obj;
+    // 4. Else,
+    //   a. Let iterator be ? Call(method, obj).
+    else
```
Use `{` if the line becomes more than 1. (even with comment).

Constellation commented on Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
```diff
@@ -0,0 +1,86 @@
+    // 5. Set wrapper.[[Iterated]] to iteratorRecord.
+    var wrapper = @wrapForValidIteratorCreate(iterator, nextMethod);
+    // 6. Return wrapper.
+    return wrapper;
```
Just return without having `wrapper` variable.

Constellation commented on Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
```diff
@@ -0,0 +1,86 @@
+    var hasInstance = @instanceOf(iterator, @Iterator.prototype);
+    // 3. If hasInstance is true, then
+    //   a. Return iteratorRecord.[[Iterator]].
+    if (hasInstance)
```
Let's just do,
```
if (@instanceOf(iterator, @Iterator.prototype))
    return iterator;
```

Constellation commented on Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
```diff
@@ -0,0 +1,86 @@
+    // 3. If method is undefined, then
+    //   a. Let iterator be obj.
+    var iterator;
+    if (@isUndefinedOrNull(method))
```
Do,
```
if (@isUndefinedOrNull(obj.@@iterator))
```

sosukesuzuki replied:
`obj.@@iterator` is used twice, so accepting your suggestion would break the following test case:

```js
{
    let symbolIteraterGetCount = 0;
    const iter = {
        get [Symbol.iterator]() {
          symbolIteraterGetCount++;
          return function() { return this };
        },
    };
    const wrapper = Iterator.from(iter);
    sameValue(symbolIteraterGetCount, 1);
}
```

So, this comment alone has not been addressed.

Constellation commented on Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
```diff
@@ -0,0 +1,86 @@
+    // 2. Let iteratorRecord be Record { [[Iterator]]: obj, [[NextMethod]]: nextMethod, [[Done]]: false }.
+    // 3. Return iteratorRecord.
+    return [obj, nextMethod];
+}
```
This function is simple. Let's not have that separately, and inline the code into the caller.

Constellation commented on Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
```diff
@@ -0,0 +1,86 @@
+    if (!@isObject(iterator))
+        @throwTypeError("bar")
+    // 6. Return ? GetIteratorDirect(iterator).
+    return @getIteratorDirect(iterator);
```
Let's expand the code used in getIteratorDirect

Constellation commented on Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
```diff
@@ -0,0 +1,86 @@
+    "use strict";
+
+    // 1. Let iteratorRecord be ? GetIteratorFlattenable(O, iterate-strings).
+    var [iterator, nextMethod] = @getIteratorFlattenable(o);
```
Let's not use destructuring assignment in builtin code.

Constellation commented on Source/JavaScriptCore/builtins/WrapForValidIteratorPrototype.js:
```diff
@@ -0,0 +1,66 @@
+    // 4. Return ? Call(iteratorRecord.[[NextMethod]], iteratorRecord.[[Iterator]]).
+    var iterator = @wrapForValidIteratorGetIteratedIterator(this);
+    var nextMethod = @wrapForValidIteratorGetIteratedNextMethod(this);
+    return nextMethod.@call(iterator);
```
Do this,

```
return @wrapForValidIteratorGetIteratedNextMethod(this).@call(@wrapForValidIteratorGetIteratedIterator(this));
```

Constellation commented on Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
```diff
@@ -0,0 +1,71 @@
+    if (!@isObject(iterator))
+        @throwTypeError("Iterator is not an object")
+    // 6. Return ? GetIteratorDirect(iterator).
+    return [iterator, iterator.next]
```
Add semicolon.

Constellation commented on Source/JavaScriptCore/builtins/JSIteratorConstructor.js:
```diff
@@ -0,0 +1,71 @@
+        iterator = method.@call(obj);
+    // 5. If iterator is not an Object, throw a TypeError exception.
+    if (!@isObject(iterator))
+        @throwTypeError("Iterator is not an object")
```
Let's add semicolon.

Constellation reviewed (approved):
r=me

sosukesuzuki committed:
[JSC] Implement `Iterator.from` from Iterator Helpers Proposal
https://bugs.webkit.org/show_bug.cgi?id=278691

Reviewed by Yusuke Suzuki.

This patch implements `Iterator.from`[2][3] from the Iterator Helpers proposal[1].

[1]: https://github.com/tc39/proposal-iterator-helpers
[2]: https://github.com/tc39/proposal-iterator-helpers?tab=readme-ov-file#iteratorfromobject
[3]: https://tc39.es/proposal-iterator-helpers/#sec-iterator.from

* JSTests/stress/iterator-from.js: Added.
(assert):
(shouldThrow):
(throw.new.Error.Iter.prototype.next):
(throw.new.Error.Iter):
(throw.new.Error):
(sameValue):
(sameValue.Iter.prototype.return):
(sameValue.Iter):
(sameValue.Iter.prototype.next):
(sameValue.const.iter.get next):
(sameValue.const.iter.get return):
(sameValue.):
(sameValue.get const):
(string_appeared_here.shouldThrow):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/JSIteratorConstructor.js: Added.
(linkTimeConstant.getIteratorDirect):
(linkTimeConstant.getIteratorFlattenable):
(from):
* Source/JavaScriptCore/builtins/WrapForValidIteratorPrototype.js: Added.
(next):
(return):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
(JSC::BytecodeGenerator::emitIsWrapForValidIterator):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/heap/HeapSubspaceTypes.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::wrapForValidIteratorPrototype const):
(JSC::JSGlobalObject::wrapForValidIteratorStructure const):
* Source/JavaScriptCore/runtime/JSIteratorConstructor.cpp:
(JSC::JSIteratorConstructor::create):
(JSC::JSIteratorConstructor::finishCreation):
* Source/JavaScriptCore/runtime/JSIteratorConstructor.h:
* Source/JavaScriptCore/runtime/JSType.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/runtime/JSType.h:
* Source/JavaScriptCore/runtime/JSWrapForValidIterator.cpp: Added.
(JSC::JSWrapForValidIterator::createWithInitialValues):
(JSC::JSWrapForValidIterator::finishCreation):
(JSC::JSWrapForValidIterator::visitChildrenImpl):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSWrapForValidIterator.h: Added.
* Source/JavaScriptCore/runtime/JSWrapForValidIteratorInlines.h:
(JSC::JSWrapForValidIterator::createStructure):
* Source/JavaScriptCore/runtime/WrapForValidIteratorPrototype.cpp:
(JSC::WrapForValidIteratorPrototype::finishCreation):
* Source/JavaScriptCore/runtime/WrapForValidIteratorPrototype.h:
* Source/JavaScriptCore/runtime/WrapForValidIteratorPrototypeInlines.h:
(JSC::WrapForValidIteratorPrototype::createStructure):

Canonical link: https://commits.webkit.org/283016@main
